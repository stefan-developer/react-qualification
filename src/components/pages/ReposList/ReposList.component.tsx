import React, { useEffect, useState } from "react";
import Page from "../../templates/Page/Page.component";
import Input from "../../atoms/Input/Input.component";
import Loader from "../../molecules/Loader/Loader.component";
import List from "../../organisms/List/List.component";

interface GithubRepo {
  name: string;
  html_url: string;
}

const getRepos = async () => {
  const reposUrl = "https://api.github.com/users/stefan-developer/repos";

  const queryOptions = {
    headers: {
      Accept: "application/vnd.github.mercy-preview+json",
    },
  };

  const response = await fetch(reposUrl, queryOptions);

  return response.json();
};

function ReposList() {
  const [loading, setLoading] = useState<boolean>(true);
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [repos, setRepos] = useState<GithubRepo[]>([]);

  useEffect(() => {
    getRepos().then((data) => {
      setRepos(data);
      setLoading(false);
    });
  }, []);

  // Logic abstracted away from presentational components. The list renderes a list and thats it.
  const filteredRepos: any = repos.filter((repo: any) =>
    repo.name.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <Page>
      <Input placeholder="Search repos" value={searchTerm} onChange={inputValue => setSearchTerm(inputValue)}/>
      {loading ? <Loader /> : <List data={filteredRepos} />}
    </Page>
  );
}

export default ReposList;
