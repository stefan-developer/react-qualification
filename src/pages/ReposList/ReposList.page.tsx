import React, { useEffect, useState } from "react";
import Page from "../../components/templates/Page/Page.component";
import Input from "../../components/atoms/Input/Input.component";
import Loader from "../../components/molecules/Loader/Loader.component";
import List from "../../components/organisms/List/List.component";
import IGithubRepo from "../../models/GithubRepo";
import getRepos from "../../api/github/get-repos/getRepos";

export function filterGithubRepos(repos: any[], searchTerm: string) {
  return repos.filter((repo: any) =>
    repo.name.toLowerCase().includes(searchTerm.toLowerCase())
  );
}

function ReposList() {
  const [loading, setLoading] = useState<boolean>(true);
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [repos, setRepos] = useState<IGithubRepo[]>([]);

  useEffect(() => {
    getRepos().then((data: IGithubRepo[]) => {
      setRepos(data);
      setLoading(false);
    });
  }, []);

  return (
    <Page>
      <Input
        placeholder="Search repos"
        value={searchTerm}
        onChange={(inputValue) => setSearchTerm(inputValue)}
      />
      {loading ? <Loader /> : <List data={filterGithubRepos(repos, searchTerm)} />}
    </Page>
  );
}

export default ReposList;
