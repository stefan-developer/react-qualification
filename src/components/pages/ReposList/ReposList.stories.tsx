import React from "react";

import { storiesOf } from "@storybook/react";

import ReposList from "./ReposList.component";

storiesOf("Page", module).add("default", () => <ReposList />);
