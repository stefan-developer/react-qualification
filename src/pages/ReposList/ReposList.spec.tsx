import React from "react";
import renderer from "react-test-renderer";
import IGithubRepo from "../../models/GithubRepo";
import ReposList, { filterGithubRepos } from "./ReposList.page";

test("ReposList renders as expected", () => {
  const component = renderer.create(<ReposList />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test("filterGithubRepos function filter the githubRepo array based on search term value", () => {
  const data = [
    { name: "Stefan", html_url: "https://github.com/stefan-developer/react-qualification" },
    { name: "github", html_url: "http://test.com" },
    { name: "repo", html_url: "http://test.com" },
    { name: "Stefan FE", html_url: "https://github.com/stefan-developer/react-qualification" }
  ] as IGithubRepo[];

  const expectedResult: IGithubRepo[] = [
    {
      name: "Stefan",
      html_url: "https://github.com/stefan-developer/react-qualification",
    },
    {
      name: "Stefan FE",
      html_url: "https://github.com/stefan-developer/react-qualification",
    },
  ];

  const actualResult = filterGithubRepos(data, "Stefan");

  expect(actualResult).toEqual(expectedResult);
});
