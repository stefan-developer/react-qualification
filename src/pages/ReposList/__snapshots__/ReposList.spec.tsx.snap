// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReposList renders as expected 1`] = `
<main
  className="sc-AxjAm iWytSx"
>
  <div
    className="sc-AxirZ hFyLEi"
  >
    <input
      onChange={[Function]}
      placeholder="Search repos"
      value=""
    />
  </div>
  <div
    className="sc-AxiKw kqNDMR"
  >
    <p>
      Loading ...
    </p>
  </div>
</main>
`;
